File: client-android/lib/src/main/kotlin/stasis/client_android/lib/discovery/ServiceDiscoveryClient.kt
package stasis.client_android.lib.discovery

import stasis.client_android.lib.utils.Try

interface ServiceDiscoveryClient : ServiceApiClient {
    val attributes: Attributes

    suspend fun latest(isInitialRequest: Boolean): Try<ServiceDiscoveryResult>

    interface Attributes {
        fun asServiceDiscoveryRequest(isInitialRequest: Boolean): ServiceDiscoveryRequest
    }
}

File: core/src/main/scala/stasis/core/discovery/ServiceDiscoveryClient.scala
package stasis.core.discovery

import scala.concurrent.Future

trait ServiceDiscoveryClient extends ServiceApiClient {
  def attributes: ServiceDiscoveryClient.Attributes
  def latest(isInitialRequest: Boolean): Future[ServiceDiscoveryResult]
}

object ServiceDiscoveryClient {
  trait Attributes { self: Product =>
    def asServiceDiscoveryRequest(isInitialRequest: Boolean): ServiceDiscoveryRequest =
      ServiceDiscoveryRequest(
        isInitialRequest = isInitialRequest,
        attributes = productIterator.zipWithIndex.map { case (value, i) =>
          productElementName(i) -> value.toString
        }.toMap
      )
  }
}
